polygon/heimdall/simulator/heimdall_simulator.go

```diff
@@ -0,0 +1,169 @@
+		activeSnapshots:    activeSnapshots,
+		knownBorSnapshots:  knownBorSnapshots,
+		activeBorSnapshots: activeBorSnapshots,
+		blockReader:        freezeblocks.NewBlockReader(activeSnapshots, activeBorSnapshots),
+		logger:             logger,
+		downloader:         downloader,
+		nextSpan:           0,
+	}
+
+	go func() {
+		<-ctx.Done()
+		s.Close()
+	}()
+
+	return s, nil
+}
+
+// FetchLatestSpan gets the next span from the snapshot
+func (h *HeimdallSimulator) FetchLatestSpan(ctx context.Context) (*heimdall.Span, error) {
+	span, err := h.getSpan(h.ctx, h.nextSpan)
+	if err != nil {
+		return nil, err
+	}
+
+	h.nextSpan++
+	return &span, nil
+}
+
+func (h *HeimdallSimulator) FetchSpan(ctx context.Context, spanID uint64) (*heimdall.Span, error) {
+	if spanID > h.nextSpan-1 {
+		return nil, errors.New("span not found")
+	}
+
+	span, err := h.getSpan(h.ctx, spanID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &span, err
+}
+
+func (h *HeimdallSimulator) FetchStateSyncEvents(ctx context.Context, fromId uint64, to time.Time, limit int) ([]*heimdall.EventRecordWithTime, error) {
+	events, err := h.getEvents(h.ctx, fromId, to, limit)
+	return events, err
+}
+
+func (h *HeimdallSimulator) FetchCheckpoint(ctx context.Context, number int64) (*heimdall.Checkpoint, error) {
+	return nil, status.Errorf(codes.Unimplemented, "method FetchCheckpoint not implemented")
+}
+
+func (h *HeimdallSimulator) FetchCheckpointCount(ctx context.Context) (int64, error) {
+	return 0, status.Errorf(codes.Unimplemented, "method FetchCheckpointCount not implemented")
+}
+
+func (h *HeimdallSimulator) FetchMilestone(ctx context.Context, number int64) (*heimdall.Milestone, error) {
+	return nil, status.Errorf(codes.Unimplemented, "method FetchMilestone not implemented")
+}
+
+func (h *HeimdallSimulator) FetchMilestoneCount(ctx context.Context) (int64, error) {
+	return 0, status.Errorf(codes.Unimplemented, "method FetchMilestoneCount not implemented")
+}
+
+func (h *HeimdallSimulator) FetchNoAckMilestone(ctx context.Context, milestoneID string) error {
+	return status.Errorf(codes.Unimplemented, "method FetchNoAckMilestone not implemented")
+}
+
+func (h *HeimdallSimulator) FetchLastNoAckMilestone(ctx context.Context) (string, error) {
+	return "", status.Errorf(codes.Unimplemented, "method FetchLastNoAckMilestone not implemented")
+}
+
+func (h *HeimdallSimulator) FetchMilestoneID(ctx context.Context, milestoneID string) error {
+	return status.Errorf(codes.Unimplemented, "method FetchMilestoneID not implemented")
+}
+
+func (h *HeimdallSimulator) Close() {
+	h.activeSnapshots.Close()
+	h.knownSnapshots.Close()
+}
+
+func (h *HeimdallSimulator) downloadData(ctx context.Context, spans *freezeblocks.Segment, sType snaptype.Type, indexFn IndexFnType) error {
+	fileName := snaptype.SegmentFileName(1, spans.From(), spans.To(), sType.Enum())
+
+	h.logger.Warn(fmt.Sprintf("Downloading %s", fileName))
+
+	err := h.downloader.Download(ctx, fileName)
```

Comment:
so I think before we Download we may want to have a check - if the .seg file for this torrent file is already downloaded & it has a generated .idx file for it then return early - no need to call Download again and Index again

Reply:
It is unlikely that this happens in this set of code, but I have added it in in case more people use this download function later.